From f46cb38dd344899857531749e4d8d1b3d7be398f Mon Sep 17 00:00:00 2001
From: Stelios Tsampas <loathingkernel@gmail.com>
Date: Tue, 17 Mar 2026 14:19:01 +0200
Subject: [PATCH] utilities: keep DXVK_FRAME_RATE and VKD3D_FRAME_RATE in the
 proton environment

---
 utilities.py | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/utilities.py b/utilities.py
index daaf190..a5b7cb4 100644
--- a/utilities.py
+++ b/utilities.py
@@ -53,9 +53,9 @@ def setup_frame_rate(env: dict, func: Callable[[dict, str, str, str], None]) ->
 
     usage: setup_frame_rate(g_session,env. prepend_to_env_str)
     """
-    frame_rate = env.pop(
+    frame_rate = env.get(
         'DXVK_FRAME_RATE',
-        env.pop('VKD3D_FRAME_RATE', env.pop('PROTON_FRAME_RATE', None)),
+        env.get('VKD3D_FRAME_RATE', env.pop('PROTON_FRAME_RATE', None)),
     )
     if frame_rate is not None:
         func(env, 'DXVK_CONFIG', f'dxgi.maxFrameRate={frame_rate}', ';')
-- 
2.53.0

