From b272091f9f43091ae054921c43933e25a27c2c82 Mon Sep 17 00:00:00 2001
From: Nikolay Sivov <nsivov@codeweavers.com>
Date: Sat, 7 Feb 2026 01:30:01 +0100
Subject: [PATCH 06/63] d2d1: Correct arc rotation angles for ellipse segments.

Signed-off-by: Nikolay Sivov <nsivov@codeweavers.com>
---
 dlls/d2d1/geometry.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/d2d1/geometry.c b/dlls/d2d1/geometry.c
index 07442260502..8f784ba7868 100644
--- a/dlls/d2d1/geometry.c
+++ b/dlls/d2d1/geometry.c
@@ -4794,7 +4794,7 @@ static void d2d_ellipse_geometry_stream(struct d2d_geometry *geometry, const D2D
 
     arcs[0].size.width = e->radiusX;
     arcs[0].size.height = e->radiusY;
-    arcs[0].rotationAngle = 90.0f;
+    arcs[0].rotationAngle = 0.0f;
     arcs[0].sweepDirection = D2D1_SWEEP_DIRECTION_CLOCKWISE;
     arcs[0].arcSize = D2D1_ARC_SIZE_SMALL;
     arcs[1] = arcs[2] = arcs[3] = arcs[0];
-- 
2.53.0

