From 23bc0c59e49de7c8850dfe0ceb8cc3d2414420be Mon Sep 17 00:00:00 2001
From: Piotr Caban <piotr@codeweavers.com>
Date: Sun, 16 Mar 2025 11:04:19 +0100
Subject: [PATCH 21/63] include: Add imaxdiv declaration.

---
 include/msvcrt/inttypes.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/include/msvcrt/inttypes.h b/include/msvcrt/inttypes.h
index a0a569d8adc..64a98e9f041 100644
--- a/include/msvcrt/inttypes.h
+++ b/include/msvcrt/inttypes.h
@@ -233,6 +233,7 @@ typedef struct {
 #define SCNuPTR "u"
 #endif
 
+_ACRTIMP imaxdiv_t          __cdecl imaxdiv(intmax_t, intmax_t);
 _ACRTIMP __int64            __cdecl strtoimax(const char*,char**,int);
 _ACRTIMP __int64            __cdecl strtoimax_l(const char*,char**,int,_locale_t);
 _ACRTIMP unsigned __int64   __cdecl strtoumax(const char*,char**,int);
-- 
2.53.0

