From 4decb89948432d71d5cc84979f997ee2cc5f8357 Mon Sep 17 00:00:00 2001
From: Hans Leidekker <hans@codeweavers.com>
Date: Fri, 21 Mar 2025 13:59:26 +0100
Subject: [PATCH 29/63] wpcap: Check for failure from pcap_dump_open().

---
 dlls/wpcap/unixlib.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/dlls/wpcap/unixlib.c b/dlls/wpcap/unixlib.c
index 87a37af3f96..e0675537e2c 100644
--- a/dlls/wpcap/unixlib.c
+++ b/dlls/wpcap/unixlib.c
@@ -176,8 +176,9 @@ static NTSTATUS wrap_dump_close( void *args )
 
 static NTSTATUS wrap_dump_open( void *args )
 {
-    const struct dump_open_params *params = args;
-    *params->ret_handle = (ULONG_PTR)pcap_dump_open( (pcap_t *)(ULONG_PTR)params->handle, params->name );
+    struct dump_open_params *params = args;
+    if (!(*params->ret_handle = (ULONG_PTR)pcap_dump_open( (pcap_t *)(ULONG_PTR)params->handle, params->name )))
+        return STATUS_NO_MEMORY;
     return STATUS_SUCCESS;
 }
 
-- 
2.53.0

