From adb379d9e8fcbdb52c4889fe4a1f0fbe10a558cd Mon Sep 17 00:00:00 2001
From: Hans Leidekker <hans@codeweavers.com>
Date: Fri, 21 Mar 2025 14:40:52 +0100
Subject: [PATCH 30/63] wpcap: Fix Unix call in pcap_dump_close().

---
 dlls/wpcap/wpcap.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/wpcap/wpcap.c b/dlls/wpcap/wpcap.c
index e8b6e4193f7..a3f00c65ac3 100644
--- a/dlls/wpcap/wpcap.c
+++ b/dlls/wpcap/wpcap.c
@@ -309,7 +309,7 @@ void CDECL pcap_dump_close( struct dumper *dumper )
 
     if (!dumper) return;
     params.handle = dumper->handle;
-    PCAP_CALL( dump, &params );
+    PCAP_CALL( dump_close, &params );
     free( dumper );
 }
 
-- 
2.53.0

