From 7e4d9f0c31a73e24a37e904bfc2750aba9c04d6a Mon Sep 17 00:00:00 2001
From: Hans Leidekker <hans@codeweavers.com>
Date: Fri, 21 Mar 2025 14:43:17 +0100
Subject: [PATCH 31/63] wpcap: Remove unneeded trace.

---
 dlls/wpcap/wpcap.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/dlls/wpcap/wpcap.c b/dlls/wpcap/wpcap.c
index a3f00c65ac3..26372cb9e87 100644
--- a/dlls/wpcap/wpcap.c
+++ b/dlls/wpcap/wpcap.c
@@ -345,8 +345,6 @@ struct dumper * CDECL pcap_dump_open( struct pcap *pcap, const char *filename )
         return NULL;
     }
 
-    TRACE( "unix_path %s\n", debugstr_a(params.name) );
-
     params.handle     = pcap->handle;
     params.ret_handle = &dumper->handle;
     if (PCAP_CALL( dump_open, &params ))
-- 
2.53.0

