From 77071d7630aca581b656afef0ea9c05a7c72d3f7 Mon Sep 17 00:00:00 2001
From: Hans Leidekker <hans@codeweavers.com>
Date: Fri, 21 Mar 2025 15:08:18 +0100
Subject: [PATCH 33/63] wpcap: Correct params structure in wow64_dump_open().

---
 dlls/wpcap/unixlib.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/wpcap/unixlib.c b/dlls/wpcap/unixlib.c
index 3c3353ac7f7..b2e2537647b 100644
--- a/dlls/wpcap/unixlib.c
+++ b/dlls/wpcap/unixlib.c
@@ -657,7 +657,7 @@ static NTSTATUS wow64_dump_open( void *args )
         PTR32 ret_handle;
     } const *params32 = args;
 
-    struct datalink_val_to_name_params params =
+    struct dump_open_params params =
     {
         params32->handle,
         ULongToPtr(params32->name),
-- 
2.53.0

